pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go

```diff
@@ -66,7 +68,25 @@ func (c *cloudProviderObserver) ObserveCloudProviderNames(genericListers configo
 		return existingConfig, append(errs, err)
 	}
 
+	featureGate, err := listers.FeatureGateLister().Get("cluster")
+	if errors.IsNotFound(err) {
+		recorder.Eventf("ObserveCloudProviderNames", "Optional featuregate.%s/cluster not found", configv1.GroupName)
```

Comment:
why wouldn't we return the previous config when hitting this condition?

Reply:
> This approach is only here until out-of-tree become the default option for openshift, all those errors means - proceed as usual with cloud-config formation. I don't want cascade effect if some added dependency will result in errors in untouched clusters. That is the reason for skipping multiple errors.

That hasn't clarified for me why it is better to destructively clear a configuration value as opposed to returning the best configuration we can.

Reply:
@Danil-Grigorev I think if you were to move this extra feature gate logic out into a helper function it may be easier to follow

If the featuregate doesn't exist, we don't return the existing config because we want to fallback to the existing logic to determine the config for the cloud provider. By putting all of the is external logic into a helper this can be better capture IMO

Reply:
I started sketching this out and then realised the problem about still needing to sync the configmap https://github.com/openshift/library-go/pull/895/files#r599857505

https://github.com/openshift/library-go/compare/master...JoelSpeed:observe-external

So actually, this doesn't really help all that much, but I think the extra `isCloudProviderExternal` function I added on that branch may help the understanding of the decision part at the top here